src/main/java/org/snakeyaml/engine/v2/composer/Composer.kt

```diff
@@ -0,0 +1,359 @@
+        node.setEndMark(endEvent.endMark)
+        inlineCommentsCollector.collectEvents()
+        if (!inlineCommentsCollector.isEmpty()) {
+            node.inLineComments = (inlineCommentsCollector.consume())
+        }
+        return node
+    }
+
+    /**
+     * Create [MappingNode]
+     *
+     * @param anchor - anchor if present
+     */
+    private fun composeMappingNode(anchor: Optional<Anchor>): MappingNode {
+        val startEvent = parser.next() as MappingStartEvent
+        val tag: Optional<String> = startEvent.tag
+        val nodeTag: Tag
+        val resolved: Boolean
+        if (!tag.isPresent || tag.get() == "!") {
+            nodeTag = Tag.MAP
+            resolved = true
+        } else {
+            nodeTag = Tag(tag.get())
+            resolved = false
+        }
+        val children: MutableList<NodeTuple> = ArrayList()
+        val node = MappingNode(
+            tag = nodeTag,
+            resolved = resolved,
+            value = children,
+            flowStyle = startEvent.flowStyle,
+            startMark = startEvent.startMark,
+            endMark = Optional.empty<Mark>(),
+        )
+        if (startEvent.isFlow()) {
+            node.blockComments = (blockCommentsCollector.consume())
+        }
+        anchor.ifPresent { a: Anchor ->
+            registerAnchor(a, node)
+        }
+        while (!parser.checkEvent(Event.ID.MappingEnd)) {
+            blockCommentsCollector.collectEvents()
+            if (parser.checkEvent(Event.ID.MappingEnd)) {
+                break
+            }
+            composeMappingChildren(children, node)
+        }
+        if (startEvent.isFlow()) {
+            node.inLineComments = (inlineCommentsCollector.collectEvents().consume())
+        }
+        val endEvent = parser.next()
+        node.setEndMark(endEvent.endMark)
+        inlineCommentsCollector.collectEvents()
+        if (!inlineCommentsCollector.isEmpty()) {
+            node.inLineComments = (inlineCommentsCollector.consume())
+        }
+        return node
+    }
+
+    /**
+     * Add the provided [node] to the children as the last child
+     *
+     * @param children - the list to be extended
+     * @param node - the child to the children
+     */
+    private fun composeMappingChildren(children: MutableList<NodeTuple>, node: MappingNode) {
+        val itemKey = composeKeyNode(node)
+        val itemValue = composeValueNode(node)
+        children.add(NodeTuple(itemKey, itemValue))
+    }
+
+    /**
+     * To be able to override `composeNode(node)` which is a key
+     *
+     * @param node - the source
+     * @return node
+     */
+    private fun composeKeyNode(node: MappingNode): Node = composeNode(Optional.of(node))
+
+    /**
+     * To be able to override `composeNode(node)` which is a value
+     *
+     * @param node - the source
+     * @return node
+     */
+    private fun composeValueNode(node: MappingNode): Node = composeNode(Optional.of(node))
```

Comment:
I don't understand the purpose of the `composeKeyNode()` and `composeValueNode()` functions.

I suspect this code can be simplified to 

```kotlin
    private fun composeMappingChildren(children: MutableList<NodeTuple>, node: MappingNode) {
        val itemKey = composeNode(Optional.of(node))
        val itemValue = composeNode(Optional.of(node))
        children.add(NodeTuple(itemKey, itemValue))
    }
```

especially if Optional can be replaced with nullable types.